pl_examples/loops_customisation/k_fold.py

```diff
@@ -0,0 +1,162 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+WARNING: Loop customization is in `pre-alpha release` and the API is likely to change quite a lot !
+Please, open issues with your own particular requests, so the Lightning Team can progressively converge to a great API.
+"""
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Type
+
+import numpy as np
+from sklearn.model_selection import KFold
+from torch.utils.data import Dataset
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning import seed_everything, Trainer
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.loops.base import ExternalLoop
+from pytorch_lightning.utilities import rank_zero_only
+from pytorch_lightning.utilities.boring_model import BoringDataModule, BoringModel
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+seed_everything(42)
+
+
+class SplitDataset(Dataset):
+    """SplitDataset is used to create Dataset Subset using indices.
+    Args:
+        dataset: A dataset to be splitted
+        indices: List of indices to expose from the dataset
+        use_duplicated_indices: Whether to allow duplicated indices.
+    Example::
+        split_ds = SplitDataset(dataset, indices=[10, 14, 25])
+        split_ds = SplitDataset(dataset, indices=[10, 10, 10, 14, 25], use_duplicated_indices=True)
+    """
+
+    _INTERNAL_KEYS = ("dataset", "indices", "data")
+
+    def __init__(self, dataset: Any, indices: List[int] = None, use_duplicated_indices: bool = False) -> None:
+        if indices is None:
+            indices = []
+        if not isinstance(indices, list):
+            raise MisconfigurationException("indices should be a list")
+
+        if use_duplicated_indices:
+            indices = list(indices)
+        else:
+            indices = list(np.unique(indices))
+
+        if np.max(indices) >= len(dataset) or np.min(indices) < 0:
+            raise MisconfigurationException(f"`indices` should be within [0, {len(dataset) -1}].")
+
+        self.dataset = dataset
```

Comment:
maybe also do a typing.cast of self here? I usually do a typecheck if i need to do some additional wrapping.